### DISCUSSION TITLE
Arrows anchored to hand drawn shapes

### DISCUSSION THREAD

**Discussion by alex-sherwin** [tldraw/tldraw-v1#562] [Ideas]
I've been following tldraw for quite some time and was really hoping that it would allow for arrows to be anchored to hand drawn shapes

Having them limited to just the built-in rectangle and ovals is useful, but if they could anchor to hand drawn shapes it would instantly make tldraw one of the best diagramming tools available IMHO

There are no tools out there which allow a seamless blend of pre-made shapes and hand drawn shapes that are treated equally in functionality with regard to tracking connections between them, and I feel like tldraw has the right foundation to take this to the next level

**Comment by Proful** [comment]
This seems to be very good idea.
"hand drawn shapes" In terms of tldraw paradigm can be thought as Group. But as soon as we select something else, Group is deselected. We may need to persist the group behaviour for longer.

To start with we may need to have the ability of managing shapes & groups.  Post that having "allow for arrows to be anchored" can be achieved easily I think.

**Comment by steveruizok** [comment]
We could certainly do this. Right now the best option is to group the shape and connect the arrow to the group, as @proful mentioned. But it would be straightforward to make drawn shapes "bindable". 

The logic for detecting the binding point may be complex, though. 

![image](https://user-images.githubusercontent.com/23072548/143239091-9e988a09-0a87-4e4a-bf4e-2fbd60f33aac.png)

To make the behavior fit with the other bindable shapes, we'd need to find all of the intersections between a ray and the shape as a polyline. (I believe this is already implemented in `@tldraw/intersect`). The anchor point would then be the point between the first and last intersection point; or the first intersection point if there is only one). 

![image](https://user-images.githubusercontent.com/23072548/143240313-51e627ce-6aa6-4ae6-b971-44837594d9aa.png)

Once we had the anchor point, from then on we would find the bound arrow handle's point by again intersecting a ray towards the anchor, finding the first intersection point, and then moving that point away from the anchor by the padding distance.

Other shape bindings also use "center" and snap to angles that point at that center. For a drawn shape, we find the average of all points and treat this as the center, though we'd be better off there by finding the [optical center](https://blog.mapbox.com/a-new-algorithm-for-finding-a-visual-center-of-a-polygon-7c77e6492fbc), which takes a bit more work.

**Comment by alex-sherwin** [comment]
I think the status quo here hasn't changed, right?

You can group hand-drawn items and the bind arrows to the group.  

But there's two key flaws here... One is that editing the group destroys to arrow binding.  The other is you can only group > 1 shape (which makes sense, since it's a "group", but in practice makes it a bit kludgy to try and anchor to hand-drawn items)

I still feel like this would be a killer feature for tldraw to make hand-drawn shapes a first-class bindable target for arrows.

I try to use tldraw as a sort of interactive whiteboard for screen sharing collaboration, and it *almost* works well for this use case, and I think this would really step up its game in that category.

With macOS/iPadOS side car and/or universal control, we're so close to having a completely seamless experience using a pencil for white boarding natively via tldraw in a browser.. this would make it near perfect

**Comment by steveruizok** [comment]
On its way for the next version :) https://twitter.com/steveruizok/status/1515730466980212742

**Reply by alex-sherwin** [reply]
Amazing, thank you!
